File: MCRunner/Strategy/StrategyBacktester.cs
using Bar = MCRunner.Instruments.Bar;
using MCRunner.Instruments;
using MCRunner.Orders;
using OrderCreator = MCRunner.Orders.OrderCreator;
using PowerLanguage.Strategy;
using PowerLanguage;
using System.Collections.Generic;
using System;

namespace MCRunner.Strategy
{
    public class StrategyBacktester<T> : IStrategyBacktester<T> where T : SignalObject
    {
        public T Strategy { get; private set; }
        public IStrategyPerformance StrategyInfo { get; private set; }
        public event Action<OrderInfo> OrderSent;
        private readonly StrategyRunner<T> runner;
        private readonly PlayableBars playableBars;

        public StrategyBacktester(IEnumerable<Bar> bars)
        {
            var orderCreator = new OrderCreator();
            orderCreator.OrderSent += (order) => OrderSent.SafeTrigger(order);

            var manager = new StrategyManager(orderCreator);
            StrategyInfo = manager.StrategyInfo;

            playableBars = new PlayableBars(bars);

            runner = new StrategyRunner<T>(
                playableBars.ToSingleDataStream(),
                orderCreator: orderCreator,
                strategyInfo: StrategyInfo);
            Strategy = runner.Strategy;
        }

        public void Backtest()
        {
            Backtest((bar) => { });
        }

        public void Backtest(Action action)
        {
            Backtest((bar) => action());
        }

        public void Backtest(Action<Bar> action)
        {
            runner.Create();
            runner.StartCalc();
            playableBars.Play((bar) =>
            {
                runner.CalcBar();
                action(bar);
            });
        }
    }
}

File: python/mcrunner/strategy/strategy_backtester.py
from typing import Iterable, Callable
from ..instruments import Bar, PlayableBars
from ..orders import OrderCreator, OrderInfo
from .strategy_runner import StrategyRunner
from .strategy_manager import StrategyManager
from .strategy_performance import StrategyPerformance

class StrategyBacktester:
    def __init__(self, strategy_cls, bars: Iterable[Bar]):
        self.order_creator = OrderCreator()
        self.manager = StrategyManager(self.order_creator)
        self.strategy_info = self.manager.StrategyInfo
        self.playable_bars = PlayableBars(bars)
        self.runner = StrategyRunner(strategy_cls,
                                     bars_data=self.playable_bars.to_single_data_stream(),
                                     order_creator=self.order_creator,
                                     strategy_info=self.strategy_info)
        self.Strategy = self.runner.Strategy

    def backtest(self, action: Callable[[Bar], None] | None = None):
        self.runner.create()
        self.runner.start_calc()
        def callback(bar):
            self.runner.calc_bar()
            index = self.runner.Strategy.Bars.CurrentBar
            self.manager.trigger_orders(bar, index)
            if action:
                action(bar)
        self.playable_bars.play(callback)
